batch.ts

```diff
@@ -0,0 +1,114 @@
+import {transact, getMutationID, setMutationID} from './db';
+import {gcal} from './gcal';
+import {userError, ErrorKind} from './errors';
+
+export async function batch(payload: any, auth: string) {
+    validatePayload(payload);
+
+    let infos = [];
+    for (let m of payload.mutations) {
+        await transact(async (db) => {
+            const currentMutationID = await getMutationID(db, payload.clientID);
+            const expectedMutationID = currentMutationID + 1;
+
+            if (m.id > expectedMutationID) {
+                throw userError(`mutation.id is too high - expected ${expectedMutationID}`);
+            }
+
+            if (m.id < expectedMutationID) {
+                // We already processed this, nbd - carry on to next one.
+                infos.push({
+                    id: m.id,
+                    error: 'mutation already processed',
+                });
+                return;
+            }
+
+            try {
+                await processMutation(m, auth);
+            } catch (e) {
+                if (e.kind == ErrorKind.UserError) {
+                    // UserError is permanent.
+                    // fall through to below, we will mark the tx handled.
+                    infos.push({
+                        id: m.id,
+                        error: e.toString(),
+                    });
+                } else {
+                    // Otherwise re-throw for top level handler to return proper error code.
+                    throw e;
+                }
+            }
+
+            await setMutationID(db, payload.clientID, expectedMutationID);
+        });
+    }
+    return {
+        mutationInfos: infos,
+    };
+}
+
+type Mutation = {
+    id: string
+    name: string
+    args: any
+}
+
+async function processMutation(m: Mutation, auth: string) {
+    const handlers = {
+        'nop': nop,
+        'serverError': serverError,
+        'addEvent': addEvent,
+        'updateEvent': updateEvent,
+        'deleteEvent': deleteEvent,
+    };
+    const h = handlers[m.name];
+    if (!h) {
+        throw userError(`unknown mutation: ${m.name}`);
+    }
+
+    await h(m.args, auth);
+}
+
+async function nop(data: any, auth: string) {
+    console.log('nop', data, auth);
+}
+
+async function serverError(data: any, auth: string) {
+    throw new Error('test error');
+}
+
+async function addEvent(data: any, auth: string) {
+    await gcal(`/calendars/${encodeURIComponent(data.calendarID)}/events`,
+        auth, {method: 'POST', body: JSON.stringify(data.body)});
+}
+
+async function updateEvent(data: any, auth: string) {
+    await gcal(`/calendars/${encodeURIComponent(data.calendarID)}/events/${data.eventID}`,
+        auth, {method: 'PATCH', body: JSON.stringify(data.body)});
+}
+
+async function deleteEvent(data, auth) {
+    await gcal(`/calendars/${encodeURIComponent(data.calendarID)}/events/${data.eventID}`,
+        auth, {method: 'DELETE'});
+}
+
+function validatePayload(payload: any) {
```

Comment:
Make this an assertion function:

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions

```ts
type Payload = {
  clientID: any;
  mutations: {id: number; name: string}[];
};
function validatePayload(payload: any): payload is Payload {
```